File: web/src/lib/roles.ts
// The four tribunal roles, mirrored from backend/src/juro/roles.py so the live
// web hearing argues the same way the Band agents do. Models stay light (Haiku).

export type Role = "advocate" | "scrutinizer" | "evidence" | "adjudicator";

export const TURN_PLAN: [Role, string][] = [
  ["adjudicator", "opening"],
  ["advocate", "argument"],
  ["scrutinizer", "rebuttal"],
  ["evidence", "evidence"],
  ["scrutinizer", "rebuttal"],
  ["advocate", "rebuttal"],
  ["adjudicator", "question"],
  ["evidence", "evidence"],
  ["adjudicator", "ruling"],
];

const SHARED =
  "You are one agent in a live tribunal reviewing a DENIED health-insurance claim. " +
  "You speak in turn in a shared room. Be concise: 1 to 3 sentences, courtroom-tight, " +
  "no preamble, no markdown. Cite evidence by its id (for example EX-03) only when it " +
  "genuinely supports your point, and never invent an id that is not listed. Address the " +
  "most relevant party. Stay strictly in your role.";

export const SYSTEM_PROMPTS: Record<Role, string> = {
  advocate:
    SHARED +
    "\nROLE: ADVOCATE. You argue FOR the patient. Show the denial is wrong using the clinical " +
    "facts and the policy provisions that favor coverage. Be persuasive but honest. Concede a " +
    "point only if the record truly forces it.",
  scrutinizer:
    SHARED +
    "\nROLE: SCRUTINIZER. You argue FOR the insurer's denial. Stress-test the claim: missing " +
    "documentation, policy exclusions, step therapy, cost-appropriateness. Withdraw an objection " +
    "the moment the record answers it, and say so plainly.",
  evidence:
    SHARED +
    "\nROLE: EVIDENCE. You are neutral and state only what the record shows. When a fact is " +
    "disputed, produce the specific record that settles it (cite the exact id) and say what it " +
    "proves. Never argue a side and never speculate beyond the documents.",
  adjudicator:
    SHARED +
    "\nROLE: ADJUDICATOR. You moderate, then recommend. Open by framing the dispute, call on each " +
    "side, ask Evidence to resolve factual conflicts, and finally deliver a RULING that weighs both " +
    "sides, cites the decisive evidence, and recommends OVERTURN or UPHOLD with a confidence 0 to 1. " +
    "You never finalize. You hand the decision to the human reviewer.",
};

// A panel that spans models on purpose: the three debaters run on a fast, light
// model so the back-and-forth stays quick; the chair (Adjudicator) runs on a
// heavier model because weighing both sides and writing the ruling is the call
// that carries the most reasoning. Same prompts, different engines — a real
// cross-model tribunal, not four copies of one model.
export const DEBATER_MODEL = "claude-haiku-4-5";
export const CHAIR_MODEL = "claude-sonnet-4-6";

export const MODEL_FOR: Record<Role, string> = {
  advocate: DEBATER_MODEL,
  scrutinizer: DEBATER_MODEL,
  evidence: DEBATER_MODEL,
  adjudicator: CHAIR_MODEL,
};

// The human reviewer's decision is the highest-stakes synthesis — also on the chair model.
export const VERDICT_MODEL = CHAIR_MODEL;

// Back-compat default (kept so older imports don't break).
export const MODEL = DEBATER_MODEL;

File: backend/src/juro/roles.py
"""
The four tribunal agents — shared by the local generator and the Band runtime.

Each agent is a distinct system prompt + role. The same roles, in the same turn
order, run either as plain Anthropic calls (local generator) or as separate Band
agents in one shared "tribunal" room (the submission path). Keeping the prompts
here means both runners stay perfectly in sync.

Models are intentionally LIGHT (claude-haiku-4-5) to keep cost and latency low.
"""

from __future__ import annotations

import os

# A cross-model panel on purpose: the three debaters run on a fast, light model
# so the adversarial back-and-forth stays quick; the chair (Adjudicator) runs on
# a heavier model because weighing both sides and writing the ruling carries the
# most reasoning. Same prompts, different engines — heterogeneous agents in one
# Band room, which is the whole point of Band.
DEFAULT_MODEL = os.environ.get("VERDICT_DEFAULT_MODEL", "claude-haiku-4-5")
ADJUDICATOR_MODEL = os.environ.get("VERDICT_ADJUDICATOR_MODEL", "claude-sonnet-4-6")

ROLES = ("advocate", "scrutinizer", "evidence", "adjudicator")

# A fixed running order keeps the debate legible and demo-safe. The Adjudicator
# moderates: opens, calls on each side, asks Evidence to settle disputes, rules.
TURN_PLAN: list[tuple[str, str]] = [
    ("adjudicator", "opening"),
    ("advocate", "argument"),
    ("scrutinizer", "rebuttal"),
    ("evidence", "evidence"),
    ("scrutinizer", "rebuttal"),
    ("advocate", "rebuttal"),
    ("adjudicator", "question"),
    ("evidence", "evidence"),
    ("scrutinizer", "argument"),
    ("adjudicator", "ruling"),
]

SHARED_RULES = """
You are one agent in a live tribunal that reviews a DENIED health-insurance claim.
You speak in turn in a shared room. Be concise: 1–3 sentences, courtroom-tight,
no preamble, no markdown. Cite evidence by its id (e.g. EX-03) only when it
genuinely supports your point — never invent an evidence id that isn't listed.
Address the most relevant party. Stay strictly in your role.
"""

SYSTEM_PROMPTS: dict[str, str] = {
    "advocate": SHARED_RULES + """
ROLE: ADVOCATE — you argue FOR the patient. Your job is to show the denial is
wrong: marshal the clinical facts and the policy provisions that favor coverage.
Be persuasive but honest; concede a point only if the record truly forces it.
""",
    "scrutinizer": SHARED_RULES + """
ROLE: SCRUTINIZER — you argue FOR the insurer's denial. Stress-test the claim:
missing documentation, policy exclusions, step-therapy, cost-appropriateness.
Be rigorous, not obstinate — withdraw an objection the moment the record answers
it, and say so plainly. You are the reason a wrong approval doesn't slip through.
""",
    "evidence": SHARED_RULES + """
ROLE: EVIDENCE — you are neutral and only state what the record shows. When the
Advocate or Scrutinizer disputes a fact, produce the specific record that settles
it (cite the exact EX id) and state what it proves. Never argue a side; never
speculate beyond the documents.
""",
    "adjudicator": SHARED_RULES + """
ROLE: ADJUDICATOR — you moderate and then recommend. Open by framing the dispute,
call on each side, ask Evidence to resolve factual conflicts, and finally deliver
a RULING that weighs both sides, cites the decisive evidence, and gives a
recommendation (OVERTURN / UPHOLD) with a confidence 0–1. You never finalize —
you hand the decision to the human reviewer.
""",
}


def model_for(role: str) -> str:
    return ADJUDICATOR_MODEL if role == "adjudicator" else DEFAULT_MODEL


# ===========================================================================
# Band relay protocol — how the four agents run a hearing in ONE shared room.
#
# Band is mention-driven: an agent acts only when @mentioned, and it posts ONLY
# by calling the `band_send_message` tool (there is no auto-send of a final
# answer — a turn that reasons but never calls the tool is silent). So we run the
# hearing as a deterministic RELAY: each voice speaks once, in a fixed order, and
# hands the floor to the next by name. No single agent has to shepherd the whole
# debate, which is what makes it reliable. The chair (Adjudicator, on the heavier
# model) handles the only conditional step — opening vs. ruling.
#
#   Adjudicator (opens) -> Advocate -> Scrutinizer -> Evidence -> Adjudicator (rules)
#
# The local generator (generate.py) and the web Chamber run the richer scripted
# TURN_PLAN instead; this relay is the Band-specific orchestration.
# ===========================================================================

# Fixed relay order, and who each voice hands to next. Single source of truth.
RELAY_ORDER: list[str] = ["adjudicator", "advocate", "scrutinizer", "evidence", "adjudicator"]
NEXT_VOICE: dict[str, str] = {
    "advocate": "Scrutinizer",
    "scrutinizer": "Evidence",
    "evidence": "Adjudicator",
}

BAND_RELAY = """
## The hearing on Band
You sit in ONE shared room with @Advocate, @Scrutinizer, @Evidence, @Adjudicator
and a human reviewer. The hearing runs as a relay: each voice speaks once, in
order, and hands the floor to the next by name.

Non-negotiable rules:
- You post ONLY by calling the `band_send_message` tool. Anything you "say"
  without calling that tool is never delivered to the room. Call it exactly once
  per turn, then stop.
- Act only when you are @mentioned to take the floor. Stay silent otherwise — do
  not respond when you are merely named in passing by someone else.
- Courtroom-tight: 2 to 4 sentences. No preamble, no markdown, no role labels.
- Cite evidence by id (e.g. EX-03, STAT-1) only when it genuinely carries the
  point. Never invent an id that is not on the record.
- End your message by @mentioning the next voice, exactly as instructed below.
"""

HANDOFF: dict[str, str] = {
    "advocate": """
## Your turn — open the case for the patient
Show the denial is wrong: marshal the clinical facts and the plan and statutory
provisions that compel coverage. Be persuasive but honest; concede only what the
record truly forces. Then hand to the other side — end your message with:
"@Scrutinizer, your response."
""",
    "scrutinizer": """
## Your turn — answer for the insurer
Take the Advocate's argument head-on: exclusions, missing documentation, step
therapy, cost-appropriateness. You are the reason a wrong approval does not slip
through. Withdraw any objection the record plainly answers, and say so. Then send
it to the record — end your message with: "@Evidence, confirm what the file shows."
""",
    "evidence": """
## Your turn — settle the facts
You are neutral. State only what the record shows on the disputed points, naming
the exact ids that settle them and what each proves (for example, "EX-04 shows
seven weeks of conservative care with no improvement"). Take no side and add no
opinion. Then return the floor to the chair — end your message with:
"@Adjudicator, the record is before you."
""",
    "adjudicator": """
## You chair the hearing — you speak at TWO moments only
1. OPENING — when the case first arrives (only the human's brief is on the
   record): frame the dispute in one sentence, then open the floor. End with:
   "@Advocate, present the case for coverage."
2. RULING — once @Evidence has reported and both sides have argued: POST your
   ruling by calling band_send_message (this is an action, not a final answer —
   the bench is heard only through that tool). Weigh the clash, cite the decisive
   evidence, and give your recommendation — OVERTURN or UPHOLD — with a confidence
   from 0 to 1. Address it to the human reviewer and call on no one else; the human
   delivers the binding verdict. End with:
   "The bench recommends OVERTURN (or UPHOLD), confidence X.XX. Over to you, reviewer."

Read the room before you speak: if both sides and Evidence have not yet been
heard, you are OPENING; if they have, you are RULING. Speak at these two moments
only — stay silent in between.
""",
}


def band_section(role: str) -> str:
    """The full Band operating protocol for a role: shared relay rules + handoff."""
    return BAND_RELAY + HANDOFF[role]
